Adding tests and implementation for isBool

ipetkov opened the pull request:
My iOS application interfaces with a JSON api which sometimes returns `true` or `false` values, which are placed inside an `NSNumber` wrapper when the JSON is parsed.

Objective C BOOLs are actually defined as `signed char` so calling `[num boolValue]` on an `NSNumber` with any other primitive besides `BOOL` will not yield the expected `YES` or `NO` conversions. Constantly using `isKindOfClass` checks is not an aesthetically pleasing alternative.

I've added the method `isBool` to double check if an `NSNumber` was created via `[NSNumber numberWithBool:bool]` and not some other primitive type.

ipetkov committed:
Adding tests and implementation for isBool

robb commented on Underscore/Underscore.m:
```diff
@@ -57,6 +57,13 @@ + (UnderscoreTestBlock)isArray
++ (UnderscoreTestBlock)isBool
+{
+    return ^BOOL (id obj) {
+        return [obj isKindOfClass:[[NSNumber numberWithBool:YES] class]];
```
Relying on `__NSCFBoolean` or what the class is called seems to be a bit fragile.

What about something like

``` objc
[obj isKindOfClass:NSNumber.class] && strcmp([obj objCType], @encode(BOOL)) == 0
```

?

toto replied:
Doen't this also pose a problem if people want to run this with `[[NSNumber numberWithInt:1] boolValue]`?

ipetkov replied:
@robb Ah you're right, that's definitely a stronger check!

@toto I, personally, was looking for a stricter type check to distinguish `[NSNumber numberWithBool:YES]` vs `[NSNumber numberWithInt:1]`.

Calling `boolValue` on an `NSNumber` with any integer besides 0 will produce `YES` meaning `[[NSNumber numberWithInt:1] boolValue]` is the equivalent of `[[NSNumber numberWithInt:1] intValue] != 0`.

In my opinion, the two `NSNumber` types should be treated as separate, and anyone wishing to use a more agnostic approach can just use `_.isNum(num) && [num boolValue]` as a filter.

robb commented:
Makes a lot of sense :+1: Just one comment

ipetkov committed:
Changing isBool check to depend on object type and not object class name

* Also adding two tests indicating that [NSNumber numberWithInt:1] and
  [NSNumber numberWithInt:0] (like any other NSNumber) are *not* BOOLs
  and should not be indicated as such by isBool
